# Task1

## Q1:

In [1]:
#Q1
import pandas as pd
import numpy as np
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

credit_risk = pd.read_csv("task1.csv") 
credit_risk.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 425 entries, 0 to 424
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Loan Purpose     425 non-null    object
 1   Checking         425 non-null    object
 2   Savings          425 non-null    object
 3   Months Customer  425 non-null    int64 
 4   Months Employed  425 non-null    int64 
 5   Gender           425 non-null    object
 6   Marital Status   425 non-null    object
 7   Age              425 non-null    int64 
 8   Housing          425 non-null    object
 9   Years            425 non-null    int64 
 10  Job              425 non-null    object
 11  Credit Risk      425 non-null    object
dtypes: int64(4), object(8)
memory usage: 40.0+ KB


Q1:

a.There are 425 bank customers in total.

b.Theren are 12 columns in total. Months Customer, Months Employed, Ages, and Years contain integer values. The columns contain string/text values are loan puposes, checking, savings, Gender, Marital Status, Housing, Job, and Credit Risk.

## Q2:

In [2]:
#Q2
credit_risk['Checking']=credit_risk['Checking'].str.strip("$").str.replace(",","").astype('int')
des_checking=credit_risk.sort_values('Checking', ascending=False)
top10=des_checking[0:10]
top10

Loan Purpose  Checking  Savings  Months Customer  Months Employed  \
259          New Car     19812      $0                25               37   
310          New Car     19766  $2,141                11               54   
114        Furniture     19155    $131                25               24   
248          New Car     18408    $212                13                9   
358  Small Appliance     17366      $0                16               21   
42   Small Appliance     16935    $189                37               60   
33          Business     16647    $895                16               34   
382  Small Appliance     16630      $0                11               47   
79          Business     15328      $0                25                9   
111        Furniture     13496    $650                19               20   

    Gender Marital Status  Age Housing  Years        Job Credit Risk  
259      M         Single   36     Own      2  Unskilled        High  
310      F       Divorced   47   Other      4  Unskilled        High  
114      M         Single   25     Own      2    Skilled         Low  
248      F       Divorced   35     Own      2    Skilled         Low  
358      M         Single   38   Other      4    Skilled        High  
42       M         Single   30     Own      2    Skilled         Low  
33       M         Single   25    Rent      4    Skilled         Low  
382      M         Single   26     Own      2    Skilled         Low  
79       M         Single   31     Own      4    Skilled         Low  
111      M         Single   33     Own      1  Unskilled        High

Q2:a.The value of the highest checking account balance is 19812, the difference between the highest and second-highest checking is 46.

## Q3:

In [3]:
#Q3.a
loan_purpose_lst=credit_risk['Loan Purpose'].value_counts()
loan_purpose_lst
#loan_purpose_lst.min()


Small Appliance    105
New Car            104
Furniture           85
Business            44
Used Car            40
Education           23
Repairs             12
Other                6
Large Appliance      4
Retraining           2
Name: Loan Purpose, dtype: int64

Q3:a.The small-appliance has the most number of applications, the retraining has the least number of applications.
There are 104 customers have applied for new car,
40 customers have applied for used-car loans, 
44 customers have applied for business loans, 
23 customers have applied for education loans, 
105 customers have applied for small-appliance loans,
 and 85 customers have applied for and furniture loans.
    

In [4]:
#Q3.b
int_saving=credit_risk["Savings"].str.strip("$").str.replace(",","").astype('int')
print('Mean value is', round(int_saving.mean(),2))
print('Median value is' , round(int_saving.median(),2))
print('Mean value is bigger than median value, that is',round(int_saving.mean(),2)>round(int_saving.median(),2))

Mean value is 1812.56
Median value is 596.0
Mean value is bigger than median value, that is True


Q3.b.Customers'average saving account balance is $1812.56, the mean value is higher than the median value of saving account balance.

In [5]:
#Q3.c
new_car = credit_risk.loc[credit_risk['Loan Purpose'] == 'New Car']
new_tb = new_car[["Loan Purpose","Checking","Savings"]]
new_checking= new_car['Checking']
#new_checking = new_tb.Checking.str.strip("$").str.replace(",","").astype('int')
new_saving = new_tb.Savings.str.strip("$").str.replace(",","").astype('int')
new_tb2 = pd.concat([new_saving,new_checking],axis = 1)

ans = new_tb2.loc[new_tb2.Checking < 500]
ans
print(len(ans),'customers have applied for new car loans and have checking account balances of less than $500.')
print(round(ans.Savings.mean(),2),'is the average savings account balance of these customers.')

Savings  Checking
2        389         0
6        229         0
11      3305         0
13       821         0
14       228       396
..       ...       ...
418      579         0
419     2684       193
422        0         0
423      712         0
424      912         0

[77 rows x 2 columns]

77 customers have applied for new car loans and have checking account balances of less than $500.
1350.51 is the average savings account balance of these customers.


Q3.c.There are 77 customers have applied for new car loans and have checking account balances of less than $500, 1350.51 is the average savings account balance of these customers.

## Q4:

In [6]:
#Q4
#int_checking=credit_risk["Checking"].str.strip("$").str.replace(",","").astype('int')
int_checking= credit_risk["Checking"]
credit_risk['Total Balance'] = int_saving + int_checking
credit_risk

def balance_status(a):
        if a < 250:
            return ('Low')
        elif a < 2000:
            return ('Medium')
        else:
            return ('High')
credit_risk['Balance Status'] = credit_risk['Total Balance'].apply(balance_status)

credit_risk



Loan Purpose  Checking  Savings  Months Customer  Months Employed  \
0    Small Appliance         0    $739                13               12   
1          Furniture         0  $1,230                25                0   
2            New Car         0    $389                19              119   
3          Furniture       638    $347                13               14   
4          Education       963  $4,754                40               45   
..               ...       ...      ...              ...              ...   
420  Small Appliance       497      $0                 7               51   
421        Furniture         0      $0                31               53   
422          New Car         0      $0                25              103   
423          New Car         0    $712                16                6   
424          New Car         0    $912                 7               39   

    Gender Marital Status  Age Housing  Years         Job Credit Risk  \
0        M         Single   23     Own      3   Unskilled         Low   
1        M       Divorced   32     Own      1     Skilled        High   
2        M         Single   38     Own      4  Management        High   
3        M         Single   36     Own      2   Unskilled        High   
4        M         Single   31    Rent      3     Skilled         Low   
..     ...            ...  ...     ...    ...         ...         ...   
420      M         Single   35   Other      4     Skilled         Low   
421      M         Single   30     Own      4     Skilled        High   
422      F       Divorced   28     Own      2     Skilled        High   
423      F       Divorced   28     Own      2     Skilled        High   
424      M         Single   44     Own      3  Management         Low   

     Total Balance  
0              739  
1             1230  
2              389  
3              985  
4             5717  
..             ...  
420            497  
421              0  
422              0  
423            712  
424            912  

[425 rows x 13 columns]

Loan Purpose  Checking  Savings  Months Customer  Months Employed  \
0    Small Appliance         0    $739                13               12   
1          Furniture         0  $1,230                25                0   
2            New Car         0    $389                19              119   
3          Furniture       638    $347                13               14   
4          Education       963  $4,754                40               45   
..               ...       ...      ...              ...              ...   
420  Small Appliance       497      $0                 7               51   
421        Furniture         0      $0                31               53   
422          New Car         0      $0                25              103   
423          New Car         0    $712                16                6   
424          New Car         0    $912                 7               39   

    Gender Marital Status  Age Housing  Years         Job Credit Risk  \
0        M         Single   23     Own      3   Unskilled         Low   
1        M       Divorced   32     Own      1     Skilled        High   
2        M         Single   38     Own      4  Management        High   
3        M         Single   36     Own      2   Unskilled        High   
4        M         Single   31    Rent      3     Skilled         Low   
..     ...            ...  ...     ...    ...         ...         ...   
420      M         Single   35   Other      4     Skilled         Low   
421      M         Single   30     Own      4     Skilled        High   
422      F       Divorced   28     Own      2     Skilled        High   
423      F       Divorced   28     Own      2     Skilled        High   
424      M         Single   44     Own      3  Management         Low   

     Total Balance Balance Status  
0              739         Medium  
1             1230         Medium  
2              389         Medium  
3              985         Medium  
4             5717           High  
..             ...            ...  
420            497         Medium  
421              0            Low  
422              0            Low  
423            712         Medium  
424            912         Medium  

[425 rows x 14 columns]

## Q5:

In [25]:
#Q5
credit_risk.pivot_table(values=['Age','Months Employed','Months Customer'],index= ['Marital Status','Gender'],aggfunc=np.mean)
#I afraid infomation missing in the first Pivot table, then I used a other way to get the same data result 
credit_risk.pivot_table(values=['Age','Months Employed','Months Customer'],index= ['Marital Status'],columns='Gender',aggfunc=np.mean,fill_value=0)



Age  Months Customer  Months Employed
Marital Status Gender                                             
Divorced       F       31.859259        20.637037        20.251852
               M       37.761905        26.285714        18.571429
Married        M       28.861111        19.916667        25.694444
Single         M       36.420601        24.360515        40.802575

Age            Months Customer             \
Gender                  F          M               F          M   
Marital Status                                                    
Divorced        31.859259  37.761905       20.637037  26.285714   
Married          0.000000  28.861111        0.000000  19.916667   
Single           0.000000  36.420601        0.000000  24.360515   

               Months Employed             
Gender                       F          M  
Marital Status                             
Divorced             20.251852  18.571429  
Married               0.000000  25.694444  
Single                0.000000  40.802575

## Q6:

In [8]:
#data may useful in Q6
credit_risk["Saving"]=int_saving
credit_risk.describe()
credit_risk['Balance Status'].value_counts()
credit_risk['Credit Risk'].value_counts()
credit_risk['Job'].value_counts()
credit_risk['Housing'].value_counts()
credit_risk['Gender'].value_counts()
credit_risk['Marital Status'].value_counts()
#credit_risk.sort_values('Credit Risk', ascending=True)


Checking  Months Customer  Months Employed         Age       Years  \
count    425.000000       425.000000       425.000000  425.000000  425.000000   
mean    1048.014118        22.896471        31.896471   34.397647    2.840000   
std     3147.183472        12.267599        32.259321   11.045126    1.087146   
min        0.000000         5.000000         0.000000   18.000000    1.000000   
25%        0.000000        13.000000         6.000000   26.000000    2.000000   
50%        0.000000        19.000000        20.000000   32.000000    3.000000   
75%      560.000000        28.000000        47.000000   41.000000    4.000000   
max    19812.000000        73.000000       119.000000   73.000000    4.000000   

       Total Balance        Saving  
count     425.000000    425.000000  
mean     2860.576471   1812.562353  
std      4826.993904   3597.285020  
min         0.000000      0.000000  
25%       490.000000    228.000000  
50%       836.000000    596.000000  
75%      2632.000000    921.000000  
max     32542.000000  19811.000000

Medium    258
High      114
Low        53
Name: Balance Status, dtype: int64

Low     214
High    211
Name: Credit Risk, dtype: int64

Skilled       271
Unskilled      89
Management     54
Unemployed     11
Name: Job, dtype: int64

Own      292
Rent      81
Other     52
Name: Housing, dtype: int64

M    290
F    135
Name: Gender, dtype: int64

Single      233
Divorced    156
Married      36
Name: Marital Status, dtype: int64

In [24]:
#credit_risk.pivot_table(values=['Age','Months Employed','Months Customer','Years'],index= ['Credit Risk','Balance Status','Loan Purpose'],aggfunc=np.mean)
credit_risk.pivot_table(values=['Age','Months Employed','Months Customer','Years'],index= ['Credit Risk','Balance Status','Job'],aggfunc=np.mean)
credit_risk.pivot_table(values=['Age','Months Employed','Months Customer','Years'],index= ['Credit Risk','Balance Status','Housing'],aggfunc=np.mean)


Age  Months Customer  \
Credit Risk Balance Status Job                                      
High        High           Management  50.200000        34.000000   
                           Skilled     31.357143        27.678571   
                           Unemployed  25.000000        13.000000   
                           Unskilled   36.307692        18.615385   
            Low            Management  35.666667        35.000000   
                           Skilled     33.434783        28.521739   
                           Unskilled   37.200000        20.800000   
            Medium         Management  38.700000        27.500000   
                           Skilled     31.416667        25.726190   
                           Unemployed  32.750000        19.750000   
                           Unskilled   33.040000        22.120000   
Low         High           Management  36.428571        22.857143   
                           Skilled     31.739130        23.065217   
                           Unemployed  24.000000         7.000000   
                           Unskilled   39.666667        14.166667   
            Low            Management  34.000000        35.500000   
                           Skilled     36.833333        19.083333   
                           Unemployed  46.000000        28.000000   
                           Unskilled   32.833333        14.666667   
            Medium         Management  36.882353        26.235294   
                           Skilled     35.205128        20.243590   
                           Unemployed  50.500000        28.000000   
                           Unskilled   37.464286        14.035714   

                                       Months Employed     Years  
Credit Risk Balance Status Job                                    
High        High           Management        55.200000  3.400000  
                           Skilled           20.178571  2.928571  
                           Unemployed         3.000000  1.000000  
                           Unskilled         21.076923  2.461538  
            Low            Management        65.333333  3.333333  
                           Skilled           37.000000  3.608696  
                           Unskilled         29.200000  3.400000  
            Medium         Management        39.050000  2.750000  
                           Skilled           28.750000  2.785714  
                           Unemployed         0.750000  2.500000  
                           Unskilled         22.880000  2.840000  
Low         High           Management        61.285714  3.000000  
                           Skilled           37.891304  2.804348  
                           Unemployed         2.500000  2.000000  
                           Unskilled         33.666667  3.166667  
            Low            Management        19.000000  4.000000  
                           Skilled           41.416667  3.166667  
                           Unemployed         0.000000  4.000000  
                           Unskilled         22.333333  3.166667  
            Medium         Management        39.823529  2.647059  
                           Skilled           33.833333  2.538462  
                           Unemployed         0.000000  3.000000  
                           Unskilled         32.428571  2.892857

Age  Months Customer  \
Credit Risk Balance Status Housing                               
High        High           Other    43.500000        32.625000   
                           Own      33.071429        24.785714   
                           Rent     32.000000        22.272727   
            Low            Other    38.833333        33.500000   
                           Own      35.058824        26.588235   
                           Rent     29.125000        26.500000   
            Medium         Other    43.823529        34.705882   
                           Own      32.802326        24.034884   
                           Rent     26.800000        22.866667   
Low         High           Other    36.600000        34.600000   
                           Own      33.488889        20.866667   
                           Rent     32.294118        17.235294   
            Low            Other    44.000000        37.000000   
                           Own      36.384615        21.307692   
                           Rent     32.333333         9.333333   
            Medium         Other    37.076923        26.615385   
                           Own      36.368932        18.864078   
                           Rent     32.777778        20.555556   

                                    Months Employed     Years  
Credit Risk Balance Status Housing                             
High        High           Other          30.500000  3.625000  
                           Own            24.500000  2.428571  
                           Rent           17.090909  3.181818  
            Low            Other          43.500000  4.000000  
                           Own            38.529412  3.235294  
                           Rent           34.625000  3.875000  
            Medium         Other          55.647059  3.764706  
                           Own            25.953488  2.500000  
                           Rent           19.766667  3.033333  
Low         High           Other          45.400000  3.200000  
                           Own            39.422222  2.644444  
                           Rent           34.117647  3.352941  
            Low            Other          24.666667  4.000000  
                           Own            27.384615  3.153846  
                           Rent           39.833333  3.333333  
            Medium         Other          26.153846  3.153846  
                           Own            34.737864  2.533981  
                           Rent           34.000000  3.111111

## Q6:


The main types of loans in the loan market are business loans, home loans, car loans, student loans, and personal loans. According to the table in Question 3a, the new and used car loans belong to car loans, so the ratio of customers who applied for car loans is 144/425. The education loan and retraining loan belong to the student loan, so the proportion of customers who want to have student loans is 25/425. The small and large appliances, furniture, and other loans belong to the personal loan. So around half of the customers who applied for a loan intend to apply for a personal loan. The percentage of business loans is around 10%.


As we can see, most people borrow money from the bank to buy items of small value, but only a few people borrow money to start a business. It is because the economy is not good after COVID-19. Reserve Bank of New Zealand(RBNZ) implemented quantitative easing to combat the economic impacts of COVID-19, so the economic problem appears significant from 2022. Annual CPI inflation has shot up at around 7% from early 2022 and will not decrease in the rest of 2022. In response to fighting the strong inflation, the RBNZ delivered its first Official Cash Rate(ORC) increase in over seven years in October 2021. The loan market turns cold gradually with the rise of interest rates.


The bank needs to do a multi-dimensional analysis of the customer's credit history, cash flow history, income(job and business), and age. Firstly, consider if the potential borrower has the ability to repay the requested loan based on their financial capacity. Then verify if the documents submitted by each specific borrower are accurate.
 
 
About the potential borrower, according to the information about bank customers who have applied for loans, the minimum total balance of the customers is zero, which means some people apply for a loan without any savings and checking; most of them should be high-risk clientele. 68% of customers who applied for a loan own at least one house, which means they may also have at least one home loan, so the bank should pay more attention if they can handle the loans.

From the tables above, the average age of the potential borrower, with high credit risk and high account balance, is the highest(50 years old). The minimum average age is 24 years old, which from the customers unemployed with low credit risk and high balance. In addition, the ratio of female customers to male customers is around 2 to 1. In contrast, according to the result of Question5, all female customers who applied for loans are divorced. That may be because divorced women have high financial pressures.


In conclusion, banks should carefully verify every customer who applies for a loan to decrease the risk that the borrower can not repay the loan in this tough economic winter.



# Task2

## Q1:

a.

i. The value of the property named “biography” is 'Travel | Food | Culture | Words'.

ii. 
This user have 181 followers.

iii. 
There are 12 posts are shown on this user's profile.

iv.
The type of the value is boolean.


b.

i. 
The type of value of "media" property is JSON Object.

ii. 
The type of value of "nodes" property is Array.

iii.
There are 12 elements in the array.

iv.
The type of values of property "comments" is JSON Object, and the type of values of property "likes" is also JSON Object.

In [10]:
#Q2
import json 
with open("task2.json", "r") as read_file:
    instag = json.load(read_file)
instag    

{'logging_page_id': 'profilePage_2814551',
 'user': {'blocked_by_viewer': False,
  'is_verified': False,
  'followed_by': {'count': 181},
  'username': 'modernuance',
  'biography': 'Travel | Food | Culture | Words',
  'id': '2814551',
  'has_requested_viewer': False,
  'follows': {'count': 231},
  'followed_by_viewer': False,
  'external_url_linkshimmed': 'http://l.instagram.com/?u=http%3A%2F%2Fvsco.co%2Fmodernuance&e=ATNBADCSWvQVulDxfm4ffJ4s6sVTsLco6qpohNB9WrGly0sIegAgQpyd1jRR8GA',
  'has_blocked_viewer': False,
  'country_block': False,
  'requested_by_viewer': False,
  'follows_viewer': False,
  'media': {'page_info': {'end_cursor': 'AQASQBvqMCvs1VHPneAK8Nd3k13mhZifCYhKZDKdgHNIJrbCIRWJPQIUXqsqjz23ikPHKecID5GKgPvZ_z81auuYE_2YjLnKUTnK6ejFwQPgqw',
    'has_next_page': True},
   'count': 104,
   'nodes': [{'code': 'BRELeVJjvlk',
     'comments_disabled': False,
     '__typename': 'GraphImage',
     'id': '1460342641979750756',
     'display_src': 'https://instagram.fmel2-1.fna.fbcdn.ne

In [11]:
#Q3:
def Average(com_list):
    return sum(com_list)/len(com_list)

nodes_len=len(instag['user']['media']['nodes'])

com_list=[]
for m in range(0,nodes_len):
    m=instag['user']['media']['nodes'][m]['comments']['count']
    com_list.append(m)
com_list_avg=Average(com_list)
print("The average number of comments of all the posts shown on this user’s profile is",round(com_list_avg,2))

likes_list=[]
for n in range(0,nodes_len):
    n=instag['user']['media']['nodes'][n]['likes']['count'] 
    likes_list.append(n)
sum_likes=sum(likes_list)
print("The total number of likes of all the posts shown on this user’s profile is",sum_likes)

The average number of comments of all the posts shown on this user’s profile is 6.67
The total number of likes of all the posts shown on this user’s profile is 1065


Q3:

a. The average number of comments of all the posts shown on this user’s profile is 6.67.


b. The total number of likes of all the posts shown on this user’s profile is 1065

In [12]:
#Q4
from datetime import datetime
date_list=[]
for f in range(0,nodes_len):
    f=instag['user']['media']['nodes'][f]['date']
    readable_date=datetime.utcfromtimestamp(f).strftime('%Y-%m-%d %H:%M:%S')
    date_list.append(readable_date)
    #print(readable_date)
oldest=min(date_list)
print("The earliest post on this user’s profile is",oldest)

The earliest post on this user’s profile is 2017-02-04 18:15:41


Q4.The earliest post on this user’s profile is 2017-02-04 18:15:41

In [13]:
#Q5
all_follows = instag['user']['follows']
fb_friends = {
    "facebook_friends": 
    {
        "count": 150
    }
             }
all_follows.update(fb_friends)
instag

{'logging_page_id': 'profilePage_2814551',
 'user': {'blocked_by_viewer': False,
  'is_verified': False,
  'followed_by': {'count': 181},
  'username': 'modernuance',
  'biography': 'Travel | Food | Culture | Words',
  'id': '2814551',
  'has_requested_viewer': False,
  'follows': {'count': 231, 'facebook_friends': {'count': 150}},
  'followed_by_viewer': False,
  'external_url_linkshimmed': 'http://l.instagram.com/?u=http%3A%2F%2Fvsco.co%2Fmodernuance&e=ATNBADCSWvQVulDxfm4ffJ4s6sVTsLco6qpohNB9WrGly0sIegAgQpyd1jRR8GA',
  'has_blocked_viewer': False,
  'country_block': False,
  'requested_by_viewer': False,
  'follows_viewer': False,
  'media': {'page_info': {'end_cursor': 'AQASQBvqMCvs1VHPneAK8Nd3k13mhZifCYhKZDKdgHNIJrbCIRWJPQIUXqsqjz23ikPHKecID5GKgPvZ_z81auuYE_2YjLnKUTnK6ejFwQPgqw',
    'has_next_page': True},
   'count': 104,
   'nodes': [{'code': 'BRELeVJjvlk',
     'comments_disabled': False,
     '__typename': 'GraphImage',
     'id': '1460342641979750756',
     'display_src': 'ht

In [14]:
#Q6
post_id = []
for v in range(0,nodes_len):
    v=instag['user']['media']['nodes'][v]['id']
    post_id.append(v)
#post_id

caption = []
for w in range(0,nodes_len):
    if "caption" in instag['user']['media']['nodes'][w]:
        w=instag['user']['media']['nodes'][w]['caption']
    else:
        w='null'
    caption.append(w)
#caption

date_list = []
for x in range(0,nodes_len):
    x=instag['user']['media']['nodes'][x]['date']
    date_list.append(x)
#date_list

comment_count = []
for y in range(0,nodes_len):
    y=instag['user']['media']['nodes'][y]['comments']['count']
    comment_count.append(y)
#comment_count

like_count = []
for z in range(0,nodes_len):
    z=instag['user']['media']['nodes'][z]['likes']['count']
    like_count.append(z)
#like_count

post_table = {
    'post_id': post_id,
    'caption':caption,
    'date': date_list,
    'comment_count': comment_count,
    'like_count':like_count,
}
posts_info = pd.DataFrame(post_table)
posts_info

import csv

with open('Task2_745973306.csv', 'w') as m:
    posts_info.to_csv(m, sep='\t', encoding='utf-8')
 

post_id                                            caption  \
0   1460342641979750756  If only they gave out awards for my top two sk...   
1   1458896568702572644                                         Stay lit ✨   
2   1458424874271656557                                    Food memories 🍴   
3   1457452226054062570                 Get friends who do both. #grateful   
4   1455252084856990151             Accents make the world a better place.   
5   1453097603029679841                                  Time to reflect ✨   
6   1452704284956991564                                  No guts, no glory   
7   1450961265198393908             Street photography (introvert edition)   
8   1447299989305867931                                       Yankee town.   
9   1446570802232410808                                               null   
10  1444528025130541071                               Saffron and symmetry   
11  1442942242691069450                                    Tunnel vision 👀   

          date  comment_count  like_count  
0   1488306430              3          81  
1   1488134045              5          62  
2   1488077814              7         220  
3   1487961866              9          46  
4   1487699588              4          74  
5   1487442754              3          56  
6   1487395867              9          99  
7   1487188083              7          89  
8   1486751625              9          96  
9   1486664699              7          77  
10  1486421181             14          87  
11  1486232141              3          78

# Task 3

## Q1:

a. 
The tag name of the root elament is PurchaseOrders

b. 
The root have 2 attributes.

c. 
The root element have 94 children.

d. The XPath for the item on Line 24 is 
/PurchaseOrders/Order[2]/OrderDate

## Q2:

In [15]:
# Q2
import xml.etree.ElementTree as ET
purchase_orders = ET.parse('task3.xml')

root = purchase_orders.getroot()
company = root.get('company')
print('The name of the company is', company)
#root.attrib.values()


The name of the company is ABC Manufacturing


In [16]:
# Q3.a
from datetime import datetime
date_format = "%m/%d/%y"
my_list=[]
for order in root:
        order_date = order.find('OrderDate')
        arrival_date = order.find('ArrivalDate')
        b=datetime.strptime(order_date.text,date_format)
        c=datetime.strptime(arrival_date.text,date_format)
        delivery_time=c-b
        days=delivery_time.days
        my_list.append(days)

print('The minimum delivery time of purchase orders is',min(my_list),'days.')
print('The maximum delivery time of purchase orders is',max(my_list),'days.')
print('The average delivery time of purchase orders is',round(sum(my_list)/len(my_list),2),'days.')


The minimum delivery time of purchase orders is 3 days.
The maximum delivery time of purchase orders is 30 days.
The average delivery time of purchase orders is 8.41 days.


Q3.a.The minimum delivery time of purchase orders is 3 days, the maximum delivery time of purchase orders is 30 days, and the average delivery time of purchase orders is 8.41 days.

In [17]:
#Q3.b
short_time=[s_element for s_element in my_list if s_element < 5]
short_time_count = len(short_time)
medium_time=[m_element for m_element in my_list if m_element >= 5 and m_element <= 9]
medium_time_count = len(medium_time)
long_time=[l_element for l_element in my_list if l_element > 9]
long_time_count=len(long_time)
print(short_time_count,'orders are classified as short.')
print(medium_time_count,'orders are classified as medium.')
print(long_time_count,'orders are classified as long.')    

3 orders are classified as short.
68 orders are classified as medium.
23 orders are classified as long.


Q3.b. There are 3 orders are classified as short, 68 orders are classified as medium, 23 orders are classified as long.

In [18]:
#Q4
import xml.etree.ElementTree as ET
from lxml import etree
import pandas as pd

my_list_str = [str(x) for x in my_list]
my_list_class = my_list.copy()
for i, n in enumerate(my_list_class):
    if n < 5:
        my_list_class[i] = "short"
    elif n >= 5 and n <= 9:
        my_list_class[i] = "medium"
    else:
        my_list_class[i] = "long"

purchase_orders = ET.parse('task3.xml')
root = purchase_orders.getroot()
child = ET.SubElement(root,"Order")
xmlstr = ET.tostring(root, encoding='utf8',method='xml')
doc = etree.fromstring(xmlstr)
for arrivaldate in doc.xpath('/PurchaseOrders/Order/ArrivalDate[1]'):
    parent = arrivaldate.getparent()
    parent.insert(parent.index(arrivaldate)+1, etree.XML('<DeliveryTime> </DeliveryTime>\n'))
    parent.insert(parent.index(arrivaldate)+2, 
                  etree.XML('<DeliveryTimeClassification> </DeliveryTimeClassification>\n'))
for i in range(0,94):
    delivery_time_index = doc[i][10]
    delivery_time_index.text = my_list_str[i]+' days'
    delivery_time_class_index = doc[i][11]
    delivery_time_class_index.text = my_list_class[i]
    i+=1
new_doc = etree.tostring(doc)
tree_final = ET.ElementTree(doc)    
df_cols = ["Supplier","OrderNo","ItemNo","ItemDescription","ItemCost","Quantity","CostPerOrder","APTerms","OrderDate","ArrivalDate","DeliveryTime","DeliveryTimeClassification"]
rows = []
for n1 in range(0, 94):
    sup = doc[n1][0].text
    orderno = doc[n1][1].text
    itemno = doc[n1][2].text
    itemdesc = doc[n1][3].text
    itemcost = doc[n1][4].text
    quantity = doc[n1][5].text
    cpo = doc[n1][6].text
    art = doc[n1][7].text
    orderdt = doc[n1][8].text
    arrdt = doc[n1][9].text
    delivtime = doc[n1][10].text
    delivclass = doc[n1][11].text
    rows.append({"Supplier":sup, "OrderNo":orderno, "ItemNo":itemno,"ItemDescription":itemdesc,"ItemCost":itemcost,"Quantity":quantity,"CostPerOrder":cpo,"APTerms":art,"OrderDate":orderdt,"ArrivalDate":arrdt,"DeliveryTime":delivtime,"DeliveryTimeClassification":delivclass})
out_df = pd.DataFrame(rows, columns = df_cols)
out_df.to_csv('Task3_745973306.csv')
out_df

Supplier   OrderNo ItemNo     ItemDescription ItemCost  \
0     Hulkey Fasteners  Aug11001   1122  Airframe fasteners    $4.25   
1        Alum Sheeting  Aug11002   1243  Airframe fasteners    $4.25   
2   Fast-Tie Aerospace  Aug11003   5462  Shielded Cable/ft.    $1.05   
3   Fast-Tie Aerospace  Aug11004   5462  Shielded Cable/ft.    $1.05   
4        Steelpin Inc.  Aug11005   5319  Shielded Cable/ft.    $1.10   
..                 ...       ...    ...                 ...      ...   
89        Manley Valve  Nov11006   9967         Hatch Decal    $0.85   
90  Fast-Tie Aerospace  Nov11007   7268      Pressure Gauge   $95.00   
91  Fast-Tie Aerospace  Nov11008   5689          Side Panel  $175.00   
92       Steelpin Inc.  Nov11009   5677          Side Panel  $195.00   
93        Manley Valve  Nov11010   9955          Door Decal    $0.55   

    Quantity   CostPerOrder APTerms OrderDate ArrivalDate DeliveryTime  \
0   "19,500"  "$82,875.00 "      30    8/5/11    08/13/11       8 days   
1   "10,000"  "$42,500.00 "      30    8/8/11    08/14/11       6 days   
2   "23,000"  "$24,150.00 "      30   8/10/11    08/15/11       5 days   
3   "21,500"  "$22,575.00 "      30  08/15/11    08/22/11       7 days   
4   "17,500"  "$19,250.00 "      30  08/20/11    08/31/11      11 days   
..       ...            ...     ...       ...         ...          ...   
89       550        $467.50      30   11/5/11    11/11/11       6 days   
90       105   "$9,975.00 "      30   11/5/11    11/16/11      11 days   
91       175  "$30,625.00 "      30   11/5/11    11/15/11      10 days   
92       110  "$21,450.00 "      30   11/5/11    11/17/11      12 days   
93       125         $68.75      30   11/5/11    11/10/11       5 days   

   DeliveryTimeClassification  
0                      medium  
1                      medium  
2                      medium  
3                      medium  
4                        long  
..                        ...  
89                     medium  
90                       long  
91                       long  
92                       long  
93                     medium  

[94 rows x 12 columns]